api/src/main/java/org/openmrs/api/storage/LocalStorageService.java

```diff
@@ -0,0 +1,195 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.storage;
+
+import javax.activation.MimetypesFileTypeMap;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.openmrs.api.StorageService;
+import org.openmrs.api.stream.StreamDataService;
+import org.openmrs.util.OpenmrsUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Conditional;
+
+/**
+ * Used to persist data in a local file system or volumes.
+ * <p>
+ * It is the default implementation of StorageService.
+ * 
+ * @since 2.8.0, 2.7.2, 2.6.16, 2.5.15
+ */
+@Conditional(StorageServiceCondition.class)
+@Qualifier("local")
```

Comment:
In order to introduce e.g. "s3" storage one would annotate the class with:
{code}
@Conditional(StorageServiceCondition.class)
@Qualifier("s3")
{code}
Then at runtime you would need to provide storage_type=s3 property to switch from local to s3.